### PR TITLE
Update to osapi/v1beta3

### PR DESCRIPTION
**Pull request by spadgett**

### CODE REVIEW

File: fixtures/build.js [hunk 1]
```diff
@@ -22,28 +20,32 @@ window.EXAMPLE_BUILD = {
...
     },
     "strategy": {
-      "type": "STI",
-      "stiStrategy": {
-        "builderImage": "openshift/ruby-20-centos7:latest",
-        "image": "openshift/ruby-20-centos7:latest"
+      "type": "Source",
+      "sourceStrategy": {
+        "from": {
+          "kind": "DockerImage",
```
Comment (liggitt) [inline fixtures/build.js:26]: From the API doc:

```
From is reference to an ImageStream, ImageStreamTag, or ImageStreamImage
```

@bparees can you confirm those are the only valid kinds currently? Do we have examples of all the valid formats for the `name` value for each of those kinds?

File: fixtures/build.js [hunk 2]
```diff
@@ -22,28 +20,32 @@ window.EXAMPLE_BUILD = {
...
+          "name": "openshift/ruby-20-centos7:latest"
```
Comment (liggitt) [inline fixtures/build.js:27]: I think `openshift` might have moved into a `namespace` field

File: fixtures/build.js [hunk 3]
```diff
@@ -22,28 +20,32 @@ window.EXAMPLE_BUILD = {
...
       "to": {
-        "kind": "ImageStream",
-        "name": "origin-ruby-sample"
-      },
-      "dockerImageReference": "172.30.73.77:5000/test/origin-ruby-sample",
-      "imageTag": "test/origin-ruby-sample",
-      "registry": "172.30.73.77:5000"
-    }
+        "kind": "ImageStreamTag",
+        "name": "origin-ruby-sample:latest"
```
Comment (liggitt) [inline fixtures/build.js:35]: @bparees same here... do we know all the formats to expect for kinds of `'ImageStreamImage', 'ImageStreamTag' or 'DockerImage'`?

File: fixtures/build.js [hunk 4]
```diff
@@ -22,28 +20,32 @@ window.EXAMPLE_BUILD = {
...
-  "completionTimestamp": "2015-04-13T18:11:30Z",
-  "config": {
-    "kind": "BuildConfig",
-    "namespace": "test",
-    "name": "ruby-sample-build"
+  "status": {
+    "phase": "Complete",
+    "startTimestamp": "2015-05-29T16:30:27Z",
+    "completionTimestamp": "2015-05-29T16:31:46Z",
+    "duration": 79000000000,
```
Comment (liggitt) [inline fixtures/build.js:44]: how long of a build is this?

File: fixtures/build.js (hunk 3 above)
Comment (bparees) [inline fixtures/build.js:35]: "From" can be ImageStream(deprecated), ImageStreamTag, ImageStreamImage, or DockerImage(pullspec)

"To" can be all of the above except ImageStreamImage (imagestreamimages are immutable so it doesn't make sense to push to one)

This code handles an image stream reference field for us, so it should cover all the logic you'd be interested in:
https://github.com/openshift/origin/blob/master/pkg/build/generator/generator.go#L284

File: fixtures/build.js (hunk 1 above)
Comment (bparees) [inline fixtures/build.js:26]: api doc appears to be out of date :(

Comment (liggitt) [inline fixtures/build.js:26]: > api doc appears to be out of date :(

you just crushed my soul

Comment (bparees) [inline fixtures/build.js:26]: achievement unlocked: soul crusher

it's mostly correct, just missing the "DockerImage" Kind.  Does that help?

Comment (liggitt) [inline fixtures/build.js:26]: In this instance, yes

Comment (liggitt) [inline fixtures/build.js:26]: @spadgett, I think for the describer bits, always show the kind and name, and if the namespace exists, show it as well (it won't for DockerImage, and may not for the ImageStream\* ones... I think current namespace is implied in that case). It's not that important for us to understand the format here (we're just displaying the data), but I was asking more for the overview case where we're correlating builds with deployment configs

File: fixtures/build.js (hunk 2 above)
Comment (liggitt) [inline fixtures/build.js:27]: so after all that, if this is really building from a `DockerImage`, `openshift/ruby-20-centos7:latest` is perfectly valid

File: fixtures/build.js (hunk 1 above)
Comment (bparees) [inline fixtures/build.js:26]: Soul status upgraded to "slightly wrinkled"?

File: views/build.html [hunk 5]
```diff
@@ -7,64 +7,65 @@
...
+    <dd>{{resource.spec.strategy.type}}</dd>
+  </dl>
+  <div ng-switch="resource.spec.strategy.type">
+    <div ng-switch-when="Source">
+      <!-- FIXME: Is this right? -->
+      <div ng-if="resource.spec.strategy.sourceStrategy.image">
         <dl class="dl-horizontal" style="margin-bottom: 0;">
           <dt>Builder image</dt>
-          <dd>{{resource.parameters.strategy.stiStrategy.image}}</dd>
+          <dd>{{resource.spec.strategy.sourceStrategy.image}}</dd>
```
Comment (liggitt) [inline views/build.html]: I think it's `...sourceStrategy.from`, which is the structured object reference discussed above

Comment (liggitt) [inline views/build.html]: actually, I think the `image` and `from` cases collapsed into `from`

File: views/build.html [hunk 6]
```diff
@@ -7,64 +7,65 @@
...
         </dl>
       </div>
-      <div ng-if="resource.parameters.strategy.stiStrategy.from">
+      <div ng-if="resource.spec.strategy.sourceStrategy.from.name">
```
Comment (liggitt) [inline views/build.html]: `<div ng-if="resource.spec.strategy.sourceStrategy.from">`

File: object-describer.js [hunk 7]
```diff
@@ -5,16 +5,6 @@ try { angular.module("openshiftUI") } catch(e) { angular.module("openshiftUI", [
 angular.module("openshiftUI").requires.push("kubernetesUI");
 
 angular.module('openshiftUI')
-.filter('imageRepoReference', function(){
```
Comment (liggitt) [inline object-describer.js:8]: Keep this, and make it do something close to https://github.com/openshift/origin/blob/master/pkg/build/generator/generator.go#L275-L325

```
if (objectRef.kind == "DockerImage") {
  return objectRef.name;
}
var ns = objectRef.namespace || "";
ns = ns == "" ? ns : ns + "/";
var ref = ns + objectRef.name;
return ref;
```

tag is embedded in the name now in the case of `kind==ImageStreamTag`, so I don't think we need it passed in separately

File: views/build.html [hunk 8]
```diff
@@ -7,64 +7,65 @@
...
         <dl class="dl-horizontal" style="margin-bottom: 0;">
-          <dt>Builder image repo [tag]</dt>
-          <dd>{{resource.parameters.strategy.stiStrategy.from | imageRepoReference : resource.parameters.strategy.stiStrategy.tag}}
-          </dd>                  
+          <dt>Builder image repo</dt>
+          <dd>{{resource.spec.strategy.sourceStrategy.from.name}}
```
Comment (liggitt) [inline views/build.html]: `resource.spec.strategy.sourceStrategy.from | imageRepoReference`

File: views/build.html [hunk 9]
```diff
@@ -7,64 +7,65 @@
...
+          </dd>
         </dl>
       </div>
     </div>
     <div ng-switch-when="Docker">
-      <div ng-if="resource.parameters.strategy.dockerStrategy.image">
+      <div ng-if="resource.spec.strategy.dockerStrategy.image">
```
Comment (liggitt) [inline views/build.html]: `image` is dead. long live `from`

File: views/build.html [hunk 10]
```diff
@@ -7,64 +7,65 @@
...
         <dl class="dl-horizontal" style="margin-bottom: 0;">
           <dt>Builder image</dt>
-          <dd>{{resource.parameters.strategy.dockerStrategy.image}}</dd>
+          <dd>{{resource.spec.strategy.dockerStrategy.image}}</dd>
```
Comment (liggitt) [inline views/build.html]: `resource.spec.strategy.dockerStrategy.from | imageRepoReference` as above

File: views/build.html [hunk 11]
```diff
@@ -7,64 +7,65 @@
...
         </dl>
       </div>
     </div>
     <div ng-switch-when="Custom">
       <dl class="dl-horizontal" style="margin-bottom: 0;">
         <dt>Builder image</dt>
-        <dd>{{resource.parameters.strategy.customStrategy.image}}</dd>
+        <dd>{{resource.spec.strategy.customStrategy.image}}</dd>
```
Comment (liggitt) [inline views/build.html]: more of the same

File: views/build.html [hunk 12]
```diff
@@ -7,64 +7,65 @@
...
     </div>
   </div>  
-  <div ng-if="resource.parameters.source">
-    <div ng-if="resource.parameters.source.type == 'Git'">
+  <div ng-if="resource.spec.source">
+    <div ng-if="resource.spec.source.type == 'Git'">
       <dl class="dl-horizontal" style="margin-bottom: 0;">
         <dt>Source repo</dt>
-        <dd>{{resource.parameters.source.git.uri}}</dd>
+        <dd>{{resource.spec.source.git.uri}}</dd>
```
Comment (liggitt) [inline views/build.html:50]: Add another `<dt>` for ref/branch:

```
<dt>Source branch/tag/ref</dt>
<dd>{{resource.spec.source.git.ref || 'master'}}</dd>
```

File: views/deployment-config.html [hunk 13]
```diff
@@ -9,23 +9,24 @@
...
   </dl>
   <h4>Selector</h4>
   <dl class="dl-horizontal">
-    <dt ng-repeat-start="(selectorKey, selectorValue) in resource.template.controllerTemplate.replicaSelector">{{selectorKey}}</dt>
+    <dt ng-repeat-start="(selectorKey, selectorValue) in resource.spec.selector">{{selectorKey}}</dt>
     <dd ng-repeat-end>{{selectorValue}}</dd>
-  </dl>  
-  <kubernetes-object-describe-pod-template template="resource.template.controllerTemplate.podTemplate.desiredState.manifest"></kubernetes-object-describe-pod-template>  
+  </dl>
+  <kubernetes-object-describe-pod-template template="resource.spec.template.spec"></kubernetes-object-describe-pod-template>
```
Comment (liggitt) [inline views/deployment-config.html:28]: pretty sure `spec.template` is going away in favor of `spec.templateRef` which is just an ObjectReference to a standalone pod template object (kind==PodTemplate, name=..., [namespace=...])

Comment (liggitt) [inline views/deployment-config.html:28]: @ironcladlou can you confirm that `deploymentConfig.spec.template` is going away in favor of `deploymentConfig.spec.templateRef` once v1beta1 dies?

File: views/route.html [hunk 14]
```diff
@@ -8,24 +8,24 @@
...
     <dt>Path</dt>
-    <dd>{{resource.path || 'None'}}</dd>
-    <dt>Service Name</dt>
-    <dd>{{resource.serviceName}}</dd>
+    <dd>{{resource.spec.path || 'None'}}</dd>
+    <dt>To</dt>
+    <dd>{{resource.spec.to.kind}} {{resource.spec.to.name}}</dd>
   </dl>
-  <div ng-if="resource.tls && resource.tls.termination">
+  <div ng-if="resource.spec.tls && resource.spec.tls.termination">
```
Comment (liggitt) [inline views/route.html]: `<div ng-if="resource.spec.tls.termination">`

File: views/deployment-config.html (hunk 13 above)
Comment (ironcladlou) [inline views/deployment-config.html:28]: @liggitt 

> @ironcladlou can you confirm that deploymentConfig.spec.template is going away in favor of deploymentConfig.spec.templateRef once v1beta1 dies?

We haven't planned for that. @smarterclayton, is this required? cc @abhgupta

Comment (smarterclayton) [inline views/deployment-config.html:28]: No, Kube v1 will support template only as will we.  Ref is going to be a v2 thing

Comment (liggitt) [inline views/deployment-config.html:28]: ok... an update in the API comment would be helpful

File: fixtures/build.js (hunk 4 above)
Comment (spadgett) [inline fixtures/build.js:44]: Hm, this was copied from actual data. I wonder if there's not a bug here.

File: views/route.html (hunk 14 above)
Comment (spadgett) [inline views/route.html]: Will fix.

File: views/build.html (hunk 8 above)
Comment (spadgett) [inline views/build.html]: Will fix.

File: views/build.html (hunk 9 above)
Comment (spadgett) [inline views/build.html]: Will fix.

File: views/build.html (hunk 10 above)
Comment (spadgett) [inline views/build.html]: Will fix.

**Comment by spadgett** [comment]
@liggitt OK, please have another look.

File: dist/object-describer.js [hunk 15]
```diff
@@ -5,13 +5,24 @@ try { angular.module("openshiftUI") } catch(e) { angular.module("openshiftUI", [
...
+      ns = ns + "/";
+    }
+    var kind = objectRef.kind;
+    if (kind === "ImageStreamTag" || kind === "ImageStreamImage") {
+      return ns+objectRef.name;
+    }
+    if (kind === "DockerImage") {
+      // TODO: replace with real DockerImageReference parse function
+      var name = objectRef.name;
+      name = name.substring(name.lastIndexOf("/")+1);
```
Comment (liggitt) [inline dist/object-describer.js]: apply the same fix here

File: object-describer.js [hunk 16]
```diff
@@ -5,13 +5,24 @@ try { angular.module("openshiftUI") } catch(e) { angular.module("openshiftUI", [
...
+      ns = ns + "/";
+    }
+    var kind = objectRef.kind;
+    if (kind === "ImageStreamTag" || kind === "ImageStreamImage") {
+      return ns+objectRef.name;
+    }
+    if (kind === "DockerImage") {
+      // TODO: replace with real DockerImageReference parse function
+      var name = objectRef.name;
+      name = name.substring(name.lastIndexOf("/")+1);
```
Comment (liggitt) [inline object-describer.js]: apply the same fix here

File: views/build.html [hunk 17]
```diff
@@ -7,63 +7,70 @@
...
+      <div ng-if="resource.spec.strategy.sourceStrategy.from">
         <dl class="dl-horizontal" style="margin-bottom: 0;">
-          <dt>Builder image repo [tag]</dt>
-          <dd>{{resource.parameters.strategy.stiStrategy.from | imageRepoReference : resource.parameters.strategy.stiStrategy.tag}}
-          </dd>                  
+          <dt>Kind</dt>
+          <dd>{{resource.spec.strategy.sourceStrategy.from.kind}}</dd>
+          <dt>Name</dt>
+          <dd>{{resource.spec.strategy.sourceStrategy.from | imageObjectRef : resource.metadata.namespace}}</dd>
+          <dt ng-if="resource.spec.strategy.sourceStrategy.from.namespace">Name</dt>
```
Comment (liggitt) [inline views/build.html]: Namespace

Comment (liggitt) [inline views/build.html]: Actually, if you're going to use `imageObjectRef` above, we probably don't need a separate line for kind and namespace

File: views/build.html [hunk 18]
```diff
@@ -7,63 +7,70 @@
...
-      <div ng-if="resource.parameters.strategy.dockerStrategy.image">
+      <div ng-if="resource.spec.strategy.dockerStrategy.from">
         <dl class="dl-horizontal" style="margin-bottom: 0;">
-          <dt>Builder image</dt>
-          <dd>{{resource.parameters.strategy.dockerStrategy.image}}</dd>
+          <dt>Kind</dt>
+          <dd>{{resource.spec.strategy.dockerStrategy.from.kind}}</dd>
+          <dt>Name</dt>
+          <dd>{{resource.spec.strategy.dockerStrategy.from | imageObjectRef : resource.metadata.namespace}}</dd>
+          <dt ng-if="resource.spec.strategy.dockerStrategy.from.namespace">Namespace</dt>
```
Comment (liggitt) [inline views/build.html]: same here for kind and namespace

File: dist/object-describer.js [hunk 19]
```diff
@@ -5,14 +5,14 @@ try { angular.module("openshiftUI") } catch(e) { angular.module("openshiftUI", [
...
-    var ns = objectRef.namespace || "";
-    ns = ns == "" ? ns : ns + "/";
-    var ref = ns + objectRef.name;
-    ref += " [" + tag + "]";
-    return ref;
+.filter('imageObjectRef', function(){
+  return function(objectRef, /* optional */ nsIfUnspecified){
+    var ns = objectRef.namespace || nsIfUnspecified;
+    if (ns) {
+      return ns + '/' + objectRef.name;
```
Comment (liggitt) [inline dist/object-describer.js:12]: don't want namespace on type "DockerImage"...

File: views/deployment-config.html [hunk 20]
```diff
@@ -9,23 +9,23 @@
     <dd>{{resource.metadata.creationTimestamp | date:'medium'}}</dd>
   </dl>
   <h3>Triggers</h3>
-  <dl class="dl-horizontal" ng-repeat="trigger in resource.triggers">
+  <dl class="dl-horizontal" ng-repeat="trigger in resource.spec.triggers">
     <dt>Type</dt>
     <dd>{{trigger.type}}</dd>
     <dt ng-if-start="trigger.type == 'ImageChange'">Image</dt>
```
Comment (liggitt) [inline views/deployment-config.html]: `&& trigger.imageChangeParams.from`

**Pull request merged**